### PR TITLE
Resgroup alter memory capabilities

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
Now we support the resgroup memory_limit alteration in below syntax:

        ALTER RESOURCE GROUP <group> SET MEMORY_LIMIT <value>;
        ALTER RESOURCE GROUP <group> SET MEMORY_SHARED_QUOTA <value>;

The new settings may take effect immediately if the actual memory usage is lower than the new settings; otherwise they will delay the effect.

### CODE REVIEW

File: src/backend/commands/resgroupcmds.c [hunk 1]
```diff
@@ -784,6 +785,33 @@ getResgroupOptionType(const char* defname)
...
+ *
+ * @return the name of type
+ */
+static const char *
+getResgroupOptionName(ResGroupLimitType type)
+{
+	switch (type)
+	{
+		case RESGROUP_LIMIT_TYPE_CONCURRENCY:
+			return "concurrency";
```
Comment (xiong-gang) [inline src/backend/commands/resgroupcmds.c]: since the constant strings appeared multiple times in the code, i would suggest we define some macros

Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: How about postpone this change to a separate PR, in fact I have some pending PRs to refactor the resgroup source code to remove duplicated logic, reduce magic numbers/strings, etc..

File: src/test/isolation2/expected/resgroup_alter_concurrency.out [hunk 2]
```diff
@@ -113,10 +117,144 @@ rg_concurrency_test|              |<IDLE> in transaction
...
+
+21:SET ROLE role_concurrency_test;
+SET
+22:SET ROLE role_concurrency_test;
+SET
+21:BEGIN;
+BEGIN
+22&:BEGIN;  <waiting ...>
+
+SELECT rsgname,waiting_reason,current_query FROM pg_stat_activity;
```
Comment (xiong-gang) [inline src/test/isolation2/expected/resgroup_alter_concurrency.out]: it's better to add where conditions like rsgname='...', or the pg_stat_activity may have some unexpected outputs

File: src/backend/commands/resgroupcmds.c [hunk 3]
```diff
@@ -601,6 +619,28 @@ AlterResourceGroup(AlterResourceGroupStmt *stmt)
 			needDispatch = true;
 			break;
 
+		case RESGROUP_LIMIT_TYPE_MEMORY_SHARED_QUOTA:
+			memSpillRatioProposed = caps.memSpillRatio.proposed;
+			if (memSpillRatioProposed + memSharedQuotaNew > RESGROUP_MAX_MEMORY_LIMIT)
```
Comment (xiong-gang) [inline src/backend/commands/resgroupcmds.c]: why is this?

File: src/backend/utils/resgroup/resgroup.c [hunk 4]
```diff
@@ -919,6 +919,75 @@ CalcMemSharedQuotaValue(int groupId, ResGroupCaps *caps, int newProposed)
...
+	memLimit = segmentChunks * caps->memLimit.proposed / 100;
+	memLimitPercent = group->memUsage * 100 / memLimit;
+	memSharedQuotaPercent = group->memSharedUsage * 100 / memLimit;
+
+	memLimitVal = caps->memLimit.value;
+	memLimitProposed = caps->memLimit.proposed;
+	memSharedQuotaVal = caps->memSharedQuota.value;
+	memSharedQuotaProposed = caps->memSharedQuota.proposed;
+
+	if (memSharedQuotaPercent > memSharedQuotaProposed)
```
Comment (xiong-gang) [inline src/backend/utils/resgroup/resgroup.c]: please add some comments here

File: src/test/isolation2/expected/resgroup_alter_concurrency.out (hunk 2 above)
Comment (pivotal-ning-yu) [inline src/test/isolation2/expected/resgroup_alter_concurrency.out]: reasonable.

File: src/backend/commands/resgroupcmds.c (hunk 3 above)
Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: As we require `memory_shared_quota + memory_spill_ratio <= 100%`, so we check it here after we have loaded the group's caps.

File: src/backend/utils/resgroup/resgroup.c (hunk 4 above)
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: sure, will add them

File: src/backend/commands/resgroupcmds.c [hunk 5]
```diff
@@ -1050,66 +1137,91 @@ insertResgroupCapabilities(Oid groupid,
...
+							   SnapshotNow, 1, &scankey);
 
-	MemSet(isnull, 0, sizeof(bool) * Natts_pg_resgroupcapability);
-	MemSet(repl, 0, sizeof(bool) * Natts_pg_resgroupcapability);
+	MemSet(values, 0, sizeof(values));
+	MemSet(isnull, 0, sizeof(isnull));
+	MemSet(repl, 0, sizeof(repl));
 
-	if (value != NULL)
+	while (HeapTupleIsValid(oldTuple = systable_getnext(sscan)))
```
Comment (guofengrichard) [inline src/backend/commands/resgroupcmds.c]: Will update of one type cause all the types of the resource group capabilities get updated?

Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: Yes, for example, first we alter memory limit and the value might not be updated due to high memory usage, then we alter memory shared quota at low memory usage, now both memory limit and memory shared quota's `value` column can be updated.

File: src/test/isolation2/expected/resgroup_alter_concurrency.out (hunk 2 above)
Comment (pivotal-ning-yu) [inline src/test/isolation2/expected/resgroup_alter_concurrency.out]: Updated in the new commit.

File: src/backend/utils/resgroup/resgroup.c (hunk 4 above)
Comment (pivotal-ning-yu) [inline src/backend/utils/resgroup/resgroup.c]: Ah, this part was already overwritten by a following commit in this PR, and that commit was well documented.

**Comment by pivotal-ning-yu** [comment]
Thanks for all the review, merged just now.

**Pull request merged**